packages/go_router/lib/src/misc/stacked_navigation_scaffold.dart

```diff
@@ -0,0 +1,207 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+import 'package:flutter/widgets.dart';
+
+import '../state.dart';
+
+/// Transition builder callback used by [IndexStackShell].
+///
+/// The builder is expected to return a transition powered by the provided
+/// `animation` and wrapping the provided `child`.
+///
+/// The `animation` provided to the builder always runs forward from 0.0 to 1.0.
+typedef StackedNavigationTransitionBuilder = Widget Function(
+  BuildContext context,
+  Animation<double> animation,
+  Widget child,
+);
+
+/// Builder for the scaffold of a [StackedNavigationScaffold]
+typedef StackedNavigationScaffoldBuilder = Widget Function(
+    BuildContext context,
+    int currentIndex,
+    List<StackedNavigationItemState> itemsState,
+    Widget scaffoldBody);
+
+/// Representation of a item in the stack of a [StackedNavigationScaffold]
+class StackedNavigationItem {
+  /// Constructs an [StackedNavigationItem].
+  StackedNavigationItem(
+      {required this.initialLocation, required this.navigatorKey});
+
+  /// The initial location/path
+  final String initialLocation;
+
+  /// Optional navigatorKey
+  final GlobalKey<NavigatorState> navigatorKey;
+}
+
+/// Represents the current state of a [StackedNavigationItem] in a
+/// [StackedNavigationScaffold]
+class StackedNavigationItemState {
+  /// Constructs an [StackedNavigationItemState].
+  StackedNavigationItemState(this.item);
+
+  /// The [StackedNavigationItem] this state is representing.
+  final StackedNavigationItem item;
+
+  /// The current [GoRouterState] for the navigator of this item.
+  GoRouterState? currentRouterState;
+
+  /// The [Navigator] for this item.
+  Navigator? navigator;
+
+  /// Gets the current location from the [currentRouterState] or falls back to
+  /// the initial location of the associated [item].
+  String get currentLocation => currentRouterState != null
+      ? currentRouterState!.location
+      : item.initialLocation;
```

Comment:
Nitpick: this could be `currentRouterState?.location ?? item.initialLocation` (not sure if this is in line with other code in this package)

Reply:
Good catch, much cleaner! Although I think I anyway need to change the way `currentLocation` is calculated, since it seems there is a bug (or I'm missing something) where the `RouteMatchList` isn't properly updated on `pop`. This in turn causes the `location` field in `GoRouterState` to remain unchanged.